aptos-move/framework/src/natives/cryptography/algebra/mod.rs

```diff
@@ -0,0 +1,362 @@
+        helpers::{make_safe_native, SafeNativeContext, SafeNativeError, SafeNativeResult},
+    },
+    safely_pop_arg,
+};
+use aptos_types::on_chain_config::{FeatureFlag, Features, TimedFeatures};
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use better_any::{Tid, TidAble};
+use move_core_types::language_storage::TypeTag;
+use move_vm_runtime::native_functions::NativeFunction;
+use move_vm_types::{
+    loaded_data::runtime_types::Type,
+    natives::function::{PartialVMError, StatusCode},
+    values::{Value, VectorRef},
+};
+use smallvec::{smallvec, SmallVec};
+use std::{any::Any, collections::VecDeque, hash::Hash, ops::Add, rc::Rc, sync::Arc};
+
+pub mod gas;
+
+/// Equivalent to `std::error::not_implemented(0)` in Move.
+const MOVE_ABORT_CODE_NOT_IMPLEMENTED: u64 = 0x0C_0000;
+
+/// This encodes an algebraic structure defined in `algebra_*.move`.
+#[derive(Copy, Clone, Eq, Hash, PartialEq)]
+pub enum Structure {
+    BLS12381Fr,
+}
+
+impl TryFrom<TypeTag> for Structure {
+    type Error = ();
+
+    fn try_from(value: TypeTag) -> Result<Self, Self::Error> {
+        match value.to_string().as_str() {
+            // Should match the full path to struct `Fr` in `algebra_bls12381.move`.
+            "0x1::algebra_bls12381::Fr" => Ok(Structure::BLS12381Fr),
+            _ => Err(()),
+        }
+    }
+}
+
+/// This encodes a supported serialization formats defined in `algebra_*.move`.
+#[derive(Copy, Clone, Eq, Hash, PartialEq)]
+pub enum SerializationFormat {
+    /// This refers to `format_bls12381fr_lsb()` in `algebra_bls12381.move`.
+    BLS12381FrLsb,
+}
+
+impl TryFrom<TypeTag> for SerializationFormat {
+    type Error = ();
+
+    fn try_from(value: TypeTag) -> Result<Self, Self::Error> {
+        match value.to_string().as_str() {
+            // Should match `format_bls12381fr_lsb()` in `algebra_bls12381.move`.
+            "0x1::algebra_bls12381::FrFormatLsb" => Ok(SerializationFormat::BLS12381FrLsb),
+            _ => Err(()),
+        }
+    }
+}
+
+fn feature_flag_of_single_type_basic_op(structure_opt: Option<Structure>) -> Option<FeatureFlag> {
+    match structure_opt {
+        Some(Structure::BLS12381Fr) => Some(FeatureFlag::BLS12_381_STRUCTURES),
+        _ => None,
+    }
+}
+
+fn feature_flag_of_serialization_format(
+    structure_opt: Option<Structure>,
+    format_opt: Option<SerializationFormat>,
+) -> Option<FeatureFlag> {
+    match (structure_opt, format_opt) {
+        (Some(Structure::BLS12381Fr), Some(SerializationFormat::BLS12381FrLsb)) => {
+            Some(FeatureFlag::BLS12_381_STRUCTURES)
+        },
+        _ => None,
+    }
+}
+
+#[derive(Tid, Default)]
+pub struct AlgebraContext {
+    objs: Vec<Rc<dyn Any>>,
+}
+
+impl AlgebraContext {
+    pub fn new() -> Self {
+        Self { objs: Vec::new() }
+    }
+}
+
+macro_rules! structure_from_ty_arg {
+    ($context:expr, $typ:expr) => {{
+        let type_tag = $context.type_to_type_tag($typ)?;
```

Comment:
Just a reminder that this may not be a cheap operation if input type is huge. We may need to implement the conversion more efficiently, leveraging the fact that there are only a few allowed types, with limited length.

Reply:
We don't necessarily have to address it in this PR, but should get this dealt with before making the algebra available to the public

Reply:
The input type will be an empty struct, e.g., `struct Fr;` for the type of a field of order $r$.

This should be fine, no?

Of course, people can call into our functions with a huge type, but they'll pay for it in gas.

Reply:
In other words, they'd be misusing the module and those calls will fail. All type arguments we use are of the form `struct StructName;`